packages/babel-plugin-proposal-pipeline-operator/src/smartVisitor.js

```diff
@@ -0,0 +1,36 @@
+import { types as t } from "@babel/core";
+import minimalVisitor from "./minimalVisitor";
+
+const updateTopicReferenceVisitor = {
+  PipelinePrimaryTopicReference(path) {
+    path.replaceWith(this.topicId);
+  },
+  AwaitExpression(path) {
+    throw path.buildCodeFrameError(
+      "await is not supported inside pipeline expressions yet",
+    );
+  },
+};
+
+const smartVisitor = {
+  ...minimalVisitor,
+  PipelineTopicExpression(path) {
+    path.get("expression").traverse(smartVisitor);
```

Comment:
The main goal here is dealing with nested pipelines. This was introduced before switching to running on exit. Maybe that isn't necessary any longer. I'll take a look at that.

Reply:
I'm coming back after fiddling with the code.

I believe I will be able to get rid of the subtraversal (maybe the one [on line 22](https://github.com/babel/babel/pull/9179/files/6921043b8f45abf124ae32c4e22557b3dbf2d659#diff-a4d6f2fbd24651e9f03fbc0acb332dd0R22) also) by storing a stack with some state in the visitor. But I'm not sure that is actually recommended (the handbook seems to discourage it). Please let me know if you think that is a worthline line of work.

Additionally can you please elaborate on your approach?

Reply:
If I understood it correctly, this traversal is needed to prevent parent topic expression from "stealing" topic references in nested expressions:

```js
foo |> # |> #;

// becomes

foo |> (_topic => _topic |> (_topic2 => _topic2));

// and not

foo |> (_topic => _topic |> (_topic2 => _topic));
```

This visitor does it by transforming the innter pipeline expressions first, but I am pretty sure that we can obtain the same behavior by skipping `PipelineTopicExpression` in `updateTopicReferenceVisitor`:

```js
const updateTopicReferenceVisitor = {
  PipelinePrimaryTopicReference(path) {
    path.replaceWith(this.topicId);
  },
  AwaitExpression(path) {
    throw path.buildCodeFrameError(
      "await is not supported inside pipeline expressions yet",
    );
  },
  PipelineTopicExpression(path) {
    path.skip();
  },
};
```

This probably would also fix https://github.com/babel/babel/pull/9179#discussion_r242001599, since we aren't using `smartVisitor` twice anymore.

Reply:
That is much more clear now. It is an excellent idea! I'm on my phone now. But I'll definitely update the code as soon as possible.

> Btw, I really like how smartVisitor inherits from minimalVisitor: being able to comment out [...]

I hadn't noticed that. It is really nice indeed!

Reply:
Your suggestion worked wonders. This is resolved AFAIK.